### PR TITLE
Upmerge 12.05.2021

### PR DESCRIPTION
**Pull request by nvlsianpu**
Regular upmerge.
Added Tesnsoflow and TraceRecorderSource Modules.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

**Comment by ioannisg** [comment]
Are you going to update the remaining repositories @nvlsianpu ?

**Comment by kapi-no** [comment]
@nvlsianpu, the Find My repository has recently been added to the NCS manifest. However, Find My CI addition to the general NCS pipeline is still pending and its failures are not visible yet.

Right now, the Find My CI is failing for this upmerge PR:

https://jenkins-ncs.nordicsemi.no/blue/organizations/jenkins/FMN%2FNCS-FMN-Test/detail/master/69/pipeline

Please hold off with this merge until these failures are fixed.

I will share with you the Find My PR number once this is fixed. Then you will be able to update the manifest for the Find My repo as a part of this PR and hopefully, we will have a passing build.

**Comment by nvlsianpu** [comment]
@kapi-no ^^ thanks for info. Will wait on your PR.

**Comment by mbolivar-nordic** [comment]
@nvlsianpu please refer to the oss-history logs for the detail on the failing check. This is the relevant part:

```

oss-history: cherry-pick failed: Command '['git', 'cherry-pick', '-x', '2294c9a82891a0be6e0694f5b0acd51cbc653181']' returned non-zero exit status 1.
oss-history: checking if 2294c9a82891a0be6e0694f5b0acd51cbc653181 is a redundant commit...
oss-history: running "git cherry-pick --abort" in "ncs/zephyr"
oss-history: running "git cherry-pick --keep-redundant-commits -x 2294c9a82891a0be6e0694f5b0acd51cbc653181" in "ncs/zephyr"
Auto-merging subsys/bluetooth/host/id.c
[detached HEAD 4e7e02800d] [nrf fromtree] Bluetooth: host: Fix undefined reference to ...
 Author: Joakim Andersson <joakim.andersson@nordicsemi.no>
 Date: Thu Apr 22 17:34:52 2021 +0200
oss-history: 2294c9a82891a0be6e0694f5b0acd51cbc653181 is a redundant commit; do you need to revert it before creating the [nrf mergeup] commit?
```

**Comment by mbolivar-nordic** [comment]
In particular this part:

```
oss-history: 2294c9a82891a0be6e0694f5b0acd51cbc653181 is a redundant commit; do you need to revert it before creating the [nrf mergeup] commit?
```

**Review by ioannisg — CHANGES_REQUESTED** [review]
https://github.com/nvlsianpu/fw-nrfconnect-nrf/pull/6

**Review by ioannisg — APPROVED** [review]

**Review by Damian-Nordic — APPROVED** [review]

File: subsys/bluetooth/controller/hci_internal.c [hunk 1]
```diff
@@ -28,7 +28,7 @@
 static struct
 {
 	bool occurred; /**< Set in only one execution context */
-	uint8_t raw_event[CONFIG_BT_RX_BUF_LEN];
+	uint8_t raw_event[CONFIG_BT_BUF_ACL_RX_SIZE + 4];
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/hci_internal.c:31]: This buffer is used to store command complete events, not ACL data. We therefore need to size the buffer accordingly.
```suggestion
	/* HCI_Read_Supported_Commands is the command which returns the largest command
	 * command complete event.
	 */
	uint8_t raw_event[struct(struct bt_hci_rp_read_supported_commands) + BT_HCI_EVT_HDR_SIZE];
```

File: subsys/bluetooth/controller/hci_driver.c [hunk 2]
```diff
@@ -356,7 +356,7 @@ static void recv_thread(void *p1, void *p2, void *p3)
 	ARG_UNUSED(p2);
 	ARG_UNUSED(p3);
 
-	static uint8_t hci_buffer[CONFIG_BT_RX_BUF_LEN];
+	static uint8_t hci_buffer[CONFIG_BT_BUF_ACL_RX_SIZE + 4];
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/hci_driver.c:359]: This HCI buffer is used to store incoming events _or_ ACL data. The buffer therefore needs to be sized accordingly. 
```suggestion
	static uint8_t hci_buffer[MAX(CONFIG_BT_BUF_ACL_RX_SIZE, BT_BUF_EVT_RX_SIZE)];
```

**Review by rugeGerritsen — CHANGES_REQUESTED** [review]
Buffers sizes used in the SoftDevice Controller driver need to be updated.

**Review by trond-snekvik — APPROVED** [review]

File: subsys/bluetooth/controller/hci_driver.c (hunk 2 above)
Comment (joerchan) [reply subsys/bluetooth/controller/hci_driver.c:359]: BT_BUF_RX_SIZE should be used for this purpose. It handles the maximum of the ACL and Events as well as the different header sizes.

File: subsys/bluetooth/controller/hci_internal.c (hunk 1 above)
Comment (joerchan) [reply subsys/bluetooth/controller/hci_internal.c:31]: I would suggest to use `BT_BUF_EVT_RX_SIZE` instead.

If no the above suggestion has a mistake (missing rest of command complete) and should be updated to:
`sizeof(struct bt_hci_rp_read_supported_commands) + CMD_COMPLETE_MIN_SIZE`

File: samples/bluetooth/mesh/light_switch/prj.conf [hunk 3]
```diff
@@ -19,7 +19,6 @@ CONFIG_DK_LIBRARY=y
 CONFIG_BT=y
 CONFIG_BT_COMPANY_ID=0x0059
 CONFIG_BT_DEVICE_NAME="Mesh Light Switch"
-CONFIG_BT_L2CAP_RX_MTU=69
 CONFIG_BT_L2CAP_TX_MTU=69
```
Comment (joerchan) [inline samples/bluetooth/mesh/light_switch/prj.conf:22]: Setting of TX_MTU (for Mesh samples with proxy protocol enabled) could be remove as well. This should be handled by the default now.

**Review by joerchan — CHANGES_REQUESTED** [review]

File: samples/bluetooth/mesh/light_switch/prj.conf (hunk 3 above)
Comment (joerchan) [reply samples/bluetooth/mesh/light_switch/prj.conf]: Sorry, this was meant as a general comment to all of the mesh/**/prj.conf that was touched. In all cases where RX_MTU was assigned to 69 TX_MTU could be removed as well.

It is not critical though, just more clean.

**Review by joerchan — APPROVED** [review]

**Comment by MarekPieta** [comment]
The machine learning application works fine, we just need the following fix to avoid build warnings: https://github.com/zephyrproject-rtos/zephyr/pull/35430.

I observed some build warnings related to nrf_desktop. I am investigating them.

File: samples/bluetooth/throughput/prj.conf [hunk 4]
```diff
@@ -28,18 +28,18 @@ CONFIG_BT_USER_DATA_LEN_UPDATE=y
...
 
-CONFIG_BT_RX_BUF_LEN=255
+CONFIG_BT_BUF_ACL_RX_SIZE=251
 CONFIG_BT_GATT_CLIENT=y
 CONFIG_BT_ATT_PREPARE_COUNT=2
 CONFIG_BT_CONN_TX_MAX=10
 CONFIG_BT_L2CAP_TX_BUF_COUNT=10
 CONFIG_BT_L2CAP_TX_MTU=247
-CONFIG_BT_L2CAP_RX_MTU=247
+CONFIG_BT_BUF_ACL_RX_SIZE=251
```
Comment (joerchan) [inline samples/bluetooth/throughput/prj.conf:37]: Duplicate entry of ACL_RX_SIZE.

**Review by MarekPieta — CHANGES_REQUESTED** [review]
I observe build issues related to B0 configuration in nrf_desktop:
`undefined reference to z_impl_k_mutex_lock`

I am investigating this right now

The same issue was replicated on `samples/bootloader` - build on nrf52840dk

**Comment by kapi-no** [comment]
@MarekPieta, I am also struggling with it in my application.

MCU boot requires Kernel API for mutexes, which is now conditionally compiled into the Zephyr application (only when the CONFIG_MULTITHREADING option is enabled). Unfortunately, the CONFIG_MULTITHREADING option increases memory for the MCU boot image.

It would be beneficial to restore the possibility of building MCUboot images without the CONFIG_MULTITHREADING option. Then, we can have better memory-optimized applications.

**Comment by MarekPieta** [comment]
Hi @kapi-no
Thanks for pointing this out. Enabling multithreading could work as workaround for our application's bootloader configurations too (the main drawback is increasing memory footprint).
@hakonfam, do we want to consider adding the workaround to the bootloader sample (it seems like the issue affects B0 too)?

**Comment by MarekPieta** [comment]
After turning on multithreading, the device crashes during initialization. It seems like the crash is related to B0 bootloader (so far the problem was replicated only for configurations with B0 bootloader). I am investigating this one.

**Comment by nvlsianpu** [comment]
@MarekPieta @kapi-no Fix is pending: https://github.com/tejlmand/nrfxlib/pull/2

**Comment by MarekPieta** [comment]
The reported issues are no longer observed. I will test remaining configurations and then approve.

**Comment by kapi-no** [comment]
@nvlsianpu, please use this PR in the Find My section of West manifest:

https://github.com/nrfconnect/sdk-find-my/pull/23

**Comment by MarekPieta** [comment]
I observed an additional issue related to nrf_desktop keyboard. The device works fine until first reboot. Then it's somehow stuck on initialization. The problem seems to be related with loading settings using a separate, preemptive thread. I will investigate this one.

**Review by pdunaj — CHANGES_REQUESTED** [review]
Adding premature nack to make sure this won't get in before we test pelion.

**Review by pdunaj — APPROVED** [review]
Added small change to follow changes done to openthread security config

**Comment by MarekPieta** [comment]
For nrf_desktop I observed some additional issues on nrf52kbd_nrf52832. The image size on nrf52820dongle needs to be reduced too (it not longer fits).

**Review by lemrey — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@grochu I cherry-picked https://github.com/nrfconnect/sdk-nrf/pull/4396/commits/5480791c510d372c13b2c26b197358353f2cfc45 here. Can you review it and give correction if needed

**Review by junqingzou — APPROVED** [review]
Changes in "applications/serial_lte_modem/src/slm_at_host.c" looks all fine

**Review by ktaborowski — CHANGES_REQUESTED** [review]
Please use homekit pr https://github.com/nrfconnect/sdk-homekit/pull/156

There are two build fails for openthread:
```
ncs/homekit/samples/multiprotocol/build/mcuboot/zephyr/include/generated/syscalls/kernel.h:86: undefined reference to `z_impl_k_sleep'
collect2: error: ld returned 1 exit status
```
and
```
ncs/zephyr/subsys/net/l2/openthread/openthread.c:35:10: fatal error: utils/uart.h: No such file or directory
   35 | #include <utils/uart.h>
      |          ^~~~~~~~~~~~~~
```
isn't the zephyr openthread commit missing? https://github.com/zephyrproject-rtos/zephyr/commit/39aee39cf996cb7adb3a2dc0298c428a00664936

could sb point me to related upstream/ncs changes?

**Comment by nvlsianpu** [comment]
^^ on the zephy upmerge branch:
``` console
anpu@ubuntu:~/codedev/NCS/zephyr$ git log -2900 --oneline | grep 'net: openthread: Align with the new NCP API'
39aee39cf9 net: openthread: Align with the new NCP API

```

**Comment by nvlsianpu** [comment]
@ktaborowski see also:
https://github.com/zephyrproject-rtos/zephyr/commit/b85250108cef66033cdb0bdbecea5d3e7926c62f
for the linker error you encountered.

**Comment by jhn-nordic** [comment]
Tensorflow repo is huge. Might cause some reactions?

**Comment by carlescufi** [comment]
> Tensorflow repo is huge. Might cause some reactions?

We should remove it from the `allowlist`. @nvlsianpu

**Comment by thst-nordic** [comment]
can someone fix the conflicts, so I can check if CI is working correctly.

**Comment by nvlsianpu** [comment]
>> Tensorflow repo is huge. Might cause some reactions?

> We should remove it from the allowlist. @nvlsianpu

Sure, can do.

**Comment by joerchan** [comment]
@nvlsianpu Please include this one: https://github.com/zephyrproject-rtos/zephyr/pull/35403 to resolve blocking by @MarekPieta

**Comment by gmarull** [comment]
@nvlsianpu docs are broken because of https://github.com/nrfconnect/sdk-nrf/pull/4527/commits/0e1c79b34df81663c3a1d70c0ce29a22d426a970#diff-9d77b30d3af3464ebcc55b8e16e76699b12fffb5e26f94edd23ae8f5f7966275

`CONFIG_BT_BUF_ACL_TX_COUNTS` -> `CONFIG_BT_BUF_ACL_TX_COUNT`

**Review by MarekPieta — APPROVED** [review]
All the reported issues are fixed. Thank you :)

**Review by gmarull — APPROVED** [review]

**Review by rugeGerritsen — APPROVED** [review]

**Review by ktaborowski — APPROVED** [review]
HomeKit configuration aligned with this upmerge

**Review by mariuszpos — APPROVED** [review]

**Review by lmaciejonczyk — APPROVED** [review]

**Review by shantha-14 — APPROVED** [review]
All downstream integration tests are green. Mesh and TFM are unstable for other issues. Hence approving.

**Review by KAGA164 — APPROVED** [review]

**Review by kapi-no — APPROVED** [review]

**Pull request merged**